caching-attribute-service-client/src/main/java/org/hypertrace/core/attribute/service/cachingclient/DefaultCachingAttributeClient.java

```diff
@@ -0,0 +1,87 @@
+package org.hypertrace.core.attribute.service.cachingclient;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableTable;
+import com.google.common.collect.Table;
+import io.grpc.CallCredentials;
+import io.grpc.Channel;
+import io.reactivex.rxjava3.core.Single;
+import java.time.Duration;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.function.Function;
+import javax.annotation.Nonnull;
+import org.hypertrace.core.attribute.service.v1.AttributeMetadata;
+import org.hypertrace.core.attribute.service.v1.AttributeMetadataFilter;
+import org.hypertrace.core.attribute.service.v1.AttributeServiceGrpc;
+import org.hypertrace.core.attribute.service.v1.AttributeServiceGrpc.AttributeServiceStub;
+
+class DefaultCachingAttributeClient implements CachingAttributeClient {
+
+  private final AttributeServiceStub attributeServiceClient;
+  private final LoadingCache<
+          AttributeCacheContextKey, Single<Table<String, String, AttributeMetadata>>>
+      cache;
+  private final AttributeMetadataFilter attributeFilter;
+
+  DefaultCachingAttributeClient(
+      @Nonnull Channel channel,
+      @Nonnull CallCredentials credentials,
+      int maxCacheContexts,
+      @Nonnull Duration cacheExpiration,
+      @Nonnull AttributeMetadataFilter attributeFilter) {
+
+    this.attributeFilter = attributeFilter;
+    this.attributeServiceClient =
+        AttributeServiceGrpc.newStub(channel).withCallCredentials(credentials);
+    this.cache =
+        CacheBuilder.newBuilder()
+            .maximumSize(maxCacheContexts)
+            .expireAfterWrite(cacheExpiration)
+            .build(CacheLoader.from(this::loadTable));
+  }
+
+  @Override
+  public Single<AttributeMetadata> get(String scope, String key) {
+    return this.getOrInvalidate(AttributeCacheContextKey.forCurrentContext())
+        .mapOptional(table -> Optional.ofNullable(table.get(scope, key)))
+        .switchIfEmpty(Single.error(this.buildErrorForMissingAttribute(scope, key)));
+  }
+
+  @Override
+  public Single<List<AttributeMetadata>> getAll() {
+    return this.getOrInvalidate(AttributeCacheContextKey.forCurrentContext())
+        .map(table -> List.copyOf(table.values()));
+  }
+
+  private Single<Table<String, String, AttributeMetadata>> loadTable(AttributeCacheContextKey key) {
+    return key.<AttributeMetadata>streamInContext(
```

Comment:
Took me some time to wrap my head around this.
It is this vs exposing a getter for the RequestContext in AttributeCacheContextKey and directly calling `GrpcClientRequestContextUtil.executeWithHeadersContext` here.

Reply:
Yeah, just delegating the responsibility of how to run something in the key's context out of this class. The cache key itself might not be the right place for it, but I don't think it's here either. Ideally, I'd probably create a context abstraction but I think given that we already have two context objects, that seems confusing. Another option would be to push this functionality into the platform context object or one of the associated utils - I think that's feasible for call and run, but I wouldn't want to introduce an Rx dependency there for stream.

A better answer might be to create a grpc stream utils package with the other ones so it can be opt in, as I've had to copy various forms of this pattern around a couple times now.